From 05219757df5aa4d7fc9a2a47f454f9170fd68798 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Fri, 28 Jun 2024 15:57:36 +0200
Subject: [PATCH 642/954] feat(stm32mp2): use the first half of SYSRAM for BL31

At runtime, BL31 will use only the first half of the SYSRAM, the upper
half will be used for non-secure DMA LLIs. To be sure nothing from this
area is still in the cache, invalidate the upper SYSRAM before enabling
BL31 cache. BL31 should then map only first half of the SYSRAM. But it
must temporarily map the upper half read-only, as this is where we will
retrieve BL2 parameters, used to fill registers for next boot stages.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git d1d113af25c0457c40c32b63950054a8d9a44856 feat(stm32mp2): use the first half of SYSRAM for BL31]
Change-Id: If42a4208a5c7214162c762c07ab115f78a6b1b9c
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/392591
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/390448
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/bl31_plat_setup.c  | 27 ++++++++++++++++++++++++++-
 plat/st/stm32mp2/stm32mp2_private.c | 10 ++++++++++
 2 files changed, 36 insertions(+), 1 deletion(-)

diff --git a/plat/st/stm32mp2/bl31_plat_setup.c b/plat/st/stm32mp2/bl31_plat_setup.c
index 4d010e5f74..b048cab96c 100644
--- a/plat/st/stm32mp2/bl31_plat_setup.c
+++ b/plat/st/stm32mp2/bl31_plat_setup.c
@@ -30,6 +30,13 @@ void bl31_early_platform_setup2(u_register_t arg0, u_register_t arg1,
 
 	stm32mp_setup_early_console();
 
+	/*
+	 * Invalidate remaining data from second half of SYSRAM (used by BL2) as this area will
+	 * be later used as non-secure.
+	 */
+	inv_dcache_range(STM32MP_SYSRAM_BASE + STM32MP_SYSRAM_SIZE / 2U,
+			 STM32MP_SYSRAM_SIZE / 2U);
+
 	mmap_add_region(BL_CODE_BASE, BL_CODE_BASE,
 			BL_CODE_END - BL_CODE_BASE,
 			MT_CODE | MT_SECURE);
@@ -65,6 +72,17 @@ void bl31_early_platform_setup2(u_register_t arg0, u_register_t arg1,
 
 	(void)stm32mp_uart_console_setup();
 
+	/*
+	 * Map upper SYSRAM where bl_params_t are stored in BL2
+	 */
+	ret = mmap_add_dynamic_region(STM32MP_SYSRAM_BASE + STM32MP_SYSRAM_SIZE / 2U,
+				      STM32MP_SYSRAM_BASE + STM32MP_SYSRAM_SIZE / 2U,
+				      STM32MP_SYSRAM_SIZE / 2U, MT_RO_DATA | MT_SECURE);
+	if (ret < 0) {
+		ERROR("BL2 params area mapping: %d\n", ret);
+		panic();
+	}
+
 	assert(arg0 != 0UL);
 	params_from_bl2 = (bl_params_t *)arg0;
 	assert(params_from_bl2 != NULL);
@@ -102,6 +120,13 @@ void bl31_early_platform_setup2(u_register_t arg0, u_register_t arg1,
 		bl_params = bl_params->next_params_info;
 	}
 
+	ret = mmap_remove_dynamic_region(STM32MP_SYSRAM_BASE + STM32MP_SYSRAM_SIZE / 2U,
+					 STM32MP_SYSRAM_SIZE / 2U);
+	if (ret < 0) {
+		ERROR("BL2 params area unmapping: %d\n", ret);
+		panic();
+	}
+
 	/*
 	 * This stm32mp_get_soc_name() call is mandatory here to store SoC part number,
 	 * and avoid using BL31 DT at runtime to check stm32mp_is_single_core().
@@ -125,7 +150,7 @@ void bl31_plat_arch_setup(void)
 
 		/* Jump manually to BL31 warm entry point, with MMU disabled. */
 		dsbsy();
-		flush_dcache_range(STM32MP_SYSRAM_BASE, STM32MP_SYSRAM_SIZE);
+		flush_dcache_range(STM32MP_SYSRAM_BASE, STM32MP_SYSRAM_SIZE / 2U);
 		disable_mmu_el3();
 		bl31_warm_entrypoint();
 		panic();
diff --git a/plat/st/stm32mp2/stm32mp2_private.c b/plat/st/stm32mp2/stm32mp2_private.c
index 5f0a0e2873..398d732ca8 100644
--- a/plat/st/stm32mp2/stm32mp2_private.c
+++ b/plat/st/stm32mp2/stm32mp2_private.c
@@ -32,12 +32,22 @@
 					MT_SECURE | \
 					MT_EXECUTE_NEVER)
 
+#if defined(IMAGE_BL31)
+/* BL31 only uses the first half of the SYSRAM */
+#define MAP_SEC_SYSRAM	MAP_REGION_FLAT(STM32MP_SYSRAM_BASE, \
+					STM32MP_SYSRAM_SIZE / 2U, \
+					MT_MEMORY | \
+					MT_RW | \
+					MT_SECURE | \
+					MT_EXECUTE_NEVER)
+#else
 #define MAP_SEC_SYSRAM	MAP_REGION_FLAT(STM32MP_SEC_SYSRAM_BASE, \
 					STM32MP_SEC_SYSRAM_SIZE, \
 					MT_MEMORY | \
 					MT_RW | \
 					MT_SECURE | \
 					MT_EXECUTE_NEVER)
+#endif
 
 #if STM32MP_USB_PROGRAMMER && !STM32MP21
 #define MAP_SEC_DEVICE_SYSRAM	MAP_REGION_FLAT(STM32MP_SEC_DEVICE_SYSRAM_BASE, \
-- 
2.53.0

